### PR TITLE
Parse CIE header to obtain the format of FDE

### PR DESCRIPTION
**Pull request by yuyichao**
Fixes unwind info registration on 32bit and AArch64. As it turns out, LLVM use `pcrel sdata8` (signed 64bit integer offset relative to pc) on x64 in `.eh_frame` and `pcrel sdata4` (signed 32bit integer offset relative to pc) on aarch64 (and very likely other formats on other archs) so the original assumption `pcrel signed` (signed pointer size integer offset relative to pc) does not hold on all archs.

The skipping of FDE length and CIE ID is also wrong on 32bits.

This can probably be cleaned up a little. In it's current state, it allows libunwind to unwind all the way to `_start` on AArch64!!! where as it could not unwind any JIT frames before.

Unfortunately doesn't seem to help the segfault on ARM (AArch32) =(.

### CODE REVIEW

**Comment by Keno** [comment]
Cool! Thank you.

**Comment by vtjnash** [comment]
does this mean we have been (incorrectly) specifying the large code model for x64 in jl_dump_shadow, or does it just always use 64-bit pointers?

**Comment by yuyichao** [comment]
Do we parse dwarf info for `jl_dump_shadow`?

**Comment by yuyichao** [comment]
Also note that this should have no effect on x64 since it is the only platform where the original assumption is correct. (Maybe PPC64 too....)

**Comment by yuyichao** [comment]
Should be ready to go. The only limitation now AFAIK is that I don't support `eh` in the augmentation string (which is supposedly only emitted by old gcc and not LLVM jit) and only `pcrel` address mode is supported (more can be added if I can figure out (or someone else can tell me) what does those mean exactly in the context of `.eh_frame`).

File: src/debuginfo.cpp [hunk 1]
```diff
@@ -1230,29 +1347,98 @@ void RTDyldMemoryManagerUnix::registerEHFrames(uint8_t *Addr,
     // While we're at it, also record the start_ip and size,
     // which we fill in the table
     unw_table_entry *table = new unw_table_entry[nentries];
+    std::vector<uintptr_t> start_ips(nentries);
     size_t cur_entry = 0;
+    // Cache the previously parsed CIE entry so that we can support multiple
+    // CIE's (may not happen) without parsing it everytime.
+    const uint8_t *cur_cie = nullptr;
```
Comment (tkelman) [inline src/debuginfo.cpp]: any risk of this causing the gcc 4, llvm 3.3 build to regress again?

Comment (yuyichao) [inline src/debuginfo.cpp]: This is llvm 37 only. There are also other c++11 features in this patch

Comment (tkelman) [inline src/debuginfo.cpp]: ah okay, didn't see an obvious ifdef for that

**Pull request merged**